### PR TITLE
KEP-4265: Update Unmasked ProcMountType to fail validation without a pod level user namespace

### PR DESCRIPTION
**Pull request by haircommander**
fail if container uses proc mount unmasked but pod does not use user namespace

#### What type of PR is this?

/kind feature
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
In the Pod API, setting the alpha `procMount` field to `Unmasked` in a container now requires setting `spec.hostUsers=false` as well.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/4265
```

### CODE REVIEW

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -4127,13 +4127,18 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 		gracePeriod = *spec.TerminationGracePeriodSeconds
 	}
 
+	var inUserNS bool
+	if spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil && !*spec.SecurityContext.HostUsers {
```
Comment (kannon92) [inline pkg/apis/core/validation/validation.go:4131]: Should we also guard this with a feature gate?

Comment (kannon92) [inline pkg/apis/core/validation/validation.go:4131]:
```suggestion
	inUserNs := spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil && !*spec.SecurityContext.HostUsers {
```

Comment (kannon92) [reply pkg/apis/core/validation/validation.go:4131]: Did a quick search for hostUsers and we don't guard this validation with a feature gate so I think this is fine actually.

File: pkg/apis/core/validation/validation_test.go [hunk 2]
```diff
@@ -7292,17 +7292,17 @@ func TestValidateEphemeralContainers(t *testing.T) {
 	} {
 		var PodRestartPolicy core.RestartPolicy
 		PodRestartPolicy = "Never"
-		if errs := validateEphemeralContainers(ephemeralContainers, containers, initContainers, vols, nil, field.NewPath("ephemeralContainers"), PodValidationOptions{}, &PodRestartPolicy); len(errs) != 0 {
+		if errs := validateEphemeralContainers(ephemeralContainers, containers, initContainers, vols, nil, field.NewPath("ephemeralContainers"), PodValidationOptions{}, &PodRestartPolicy, false); len(errs) != 0 {
```
Comment (kannon92) [inline pkg/apis/core/validation/validation_test.go:7295]: I found bool flags for these type of functions always cause confusion in the long run. Can you add a named variable for this like:

```
noUserNamespace := false

and just pass that it in these tests? It is not clear what false means in this case. 
Even better would be having this function take a enum to make it clear what this means

File: pkg/apis/core/validation/validation.go (hunk 1 above)
Comment (haircommander) [reply pkg/apis/core/validation/validation.go:4131]: we do already do some validation on the proc mount field in this file, and there's no mention of the feature, which implies to me the field is already [filtered out](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L775) if ProcMount isn't an available option (or it doesn't matter if it's done before, as happening after works too)

Comment (kannon92) [reply pkg/apis/core/validation/validation.go:4131]: Sounds good to me.

**Comment by kannon92** [comment]
@haircommander I think you may need to look into those integration failures. Logs are unhelpful but it seems to be a consistent failure and not a flake.

**Comment by haircommander** [comment]
> @haircommander I think you may need to look into those integration failures. Logs are unhelpful but it seems to be a consistent failure and not a flake.

yeah thanks I'm looking into it, it's definitely legit

**Review by mrunalp — APPROVED** [review]

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -4139,13 +4139,15 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
...
+	inUserNS := spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil && !*spec.SecurityContext.HostUsers
+
 	vols, vErrs := ValidateVolumes(spec.Volumes, podMeta, fldPath.Child("volumes"), opts)
 	allErrs = append(allErrs, vErrs...)
 	podClaimNames := gatherPodResourceClaimNames(spec.ResourceClaims)
 	allErrs = append(allErrs, validatePodResourceClaims(podMeta, spec.ResourceClaims, fldPath.Child("resourceClaims"))...)
-	allErrs = append(allErrs, validateContainers(spec.Containers, vols, podClaimNames, gracePeriod, fldPath.Child("containers"), opts, &spec.RestartPolicy)...)
-	allErrs = append(allErrs, validateInitContainers(spec.InitContainers, spec.Containers, vols, podClaimNames, gracePeriod, fldPath.Child("initContainers"), opts, &spec.RestartPolicy)...)
-	allErrs = append(allErrs, validateEphemeralContainers(spec.EphemeralContainers, spec.Containers, spec.InitContainers, vols, podClaimNames, fldPath.Child("ephemeralContainers"), opts, &spec.RestartPolicy)...)
+	allErrs = append(allErrs, validateContainers(spec.Containers, vols, podClaimNames, gracePeriod, fldPath.Child("containers"), opts, &spec.RestartPolicy, inUserNS)...)
```
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:4148]: aside: There's a lot of options that need to be plumbed to container validation. Maybe we should just pass the full pod spec down, or create a struct to hold the relevant bits... Not something you need to do for this PR.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -4139,13 +4139,15 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 		gracePeriod = *spec.TerminationGracePeriodSeconds
 	}
 
+	inUserNS := spec.SecurityContext != nil && spec.SecurityContext.HostUsers != nil && !*spec.SecurityContext.HostUsers
```
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:4142]: nit: Consider inverting this and naming it `hostUsers`, to make it clear where the value comes from.

File: pkg/apis/core/validation/validation.go [hunk 5]
```diff
@@ -7080,6 +7082,9 @@ func ValidateSecurityContext(sc *core.SecurityContext, fldPath *field.Path) fiel
 		if err := ValidateProcMountType(fldPath.Child("procMount"), *sc.ProcMount); err != nil {
 			allErrs = append(allErrs, err)
 		}
+		if !inUserNS && sc.ProcMount != nil && *sc.ProcMount == core.UnmaskedProcMount {
```
Comment (tallclair) [inline pkg/apis/core/validation/validation.go:7085]: This is tightening validation. I'm not sure what the implications of that are for an alpha feature though. But you probably need an api reviewer to take a look (tag it with `/label api-review`). You probably need to plumb through an option to skip this validation if the object is already in an invalid state.

File: pkg/features/kube_features.go [hunk 6]
```diff
@@ -1136,7 +1137,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	PortForwardWebsockets: {Default: false, PreRelease: featuregate.Alpha},
 
-	ProcMountType: {Default: false, PreRelease: featuregate.Alpha},
+	ProcMountType: {Default: false, PreRelease: featuregate.Beta},
```
Comment (tallclair) [inline pkg/features/kube_features.go:1140]: Beta features should default to true, unless there's a very good reason not to.

**Review by tallclair — COMMENTED** [review]
No concerns from a pod-security-admission point of view - the changes are pretty trivial. I have a couple other concerns though, and this will need an API reviewer to take a look.

File: pkg/apis/core/validation/validation.go (hunk 3 above)
Comment (haircommander) [reply pkg/apis/core/validation/validation.go:4148]: yeah I agree though also agree it felt out of scope here

File: pkg/features/kube_features.go (hunk 6 above)
Comment (haircommander) [reply pkg/features/kube_features.go:1140]: I followed [userns](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go#L1204) . I'm happy to update both but I figure this should be in the same enabled state as that

File: pkg/apis/core/validation/validation.go (hunk 3 above)
Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: narrowly plumbing down the boolean seems fine for this PR

File: pkg/features/kube_features.go (hunk 6 above)
Comment (haircommander) [reply pkg/features/kube_features.go]: updated to have both enabled by default

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (haircommander) [reply pkg/apis/core/validation/validation.go]: can you explain further? I belive this field is stripped when the feature gate isn't enabled, so I'm not sure the condition we would add the pod validation option or not

File: pkg/apis/core/validation/validation.go (hunk 4 above)
Comment (haircommander) [reply pkg/apis/core/validation/validation.go]: updated!

File: pkg/features/kube_features.go (hunk 6 above)
Comment (liggitt) [reply pkg/features/kube_features.go]: Can you pull the feature gate promotion out of this PR?

The validation change in isolation looks pretty reasonable, but I need to think about the interaction with the user namespace feature (both in terms of what n-1 API servers will allow through validation once objects exist with data in this field, and in terms of what will happen if the user namespace feature gets disabled, clearing out that field while keeping this procMount field).

procMount validation relying on hostUser field data may mean we have to sequence feature graduation between the two features.

File: pkg/apis/core/validation/validation.go (hunk 5 above)
Comment (liggitt) [reply pkg/apis/core/validation/validation.go]: Changing validation while in alpha is fair game... I'm not sure I would add the complexity of ratcheting for it.

A release note is probably sufficient, but the validation change likely means we'll wait to move this to beta until 1.31

File: pkg/apis/core/validation/validation.go [hunk 7]
```diff
@@ -7080,6 +7084,9 @@ func ValidateSecurityContext(sc *core.SecurityContext, fldPath *field.Path) fiel
 		if err := ValidateProcMountType(fldPath.Child("procMount"), *sc.ProcMount); err != nil {
 			allErrs = append(allErrs, err)
 		}
+		if hostUsers && sc.ProcMount != nil && *sc.ProcMount == core.UnmaskedProcMount {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:7087]: nit: nil check is redundant

Comment (haircommander) [reply pkg/apis/core/validation/validation.go]: dropped!

File: pkg/features/kube_features.go (hunk 6 above)
Comment (haircommander) [reply pkg/features/kube_features.go]: removed it! my ideal is they graduate to beta together, though I'd like this to move forward more

> in terms of what will happen if the user namespace feature gets disabled, clearing out that field while keeping this procMount field

WDYT about an option where we clear this field if UserNamespacesSupport and ProcMountType isn't enabled?

**Review by mrunalp — APPROVED** [review]
Looks fine from node perspective.

**Comment by liggitt** [comment]
/hold

unit test failure is related:

```
{Failed;  === RUN   TestValidateSecurityContext
    validation_test.go:23028: [proc mount unmasked] Expected success, got [field.procMount: Invalid value: "Unmasked": `hostUsers` must be false to use `Unmasked`]
    validation_test.go:23087: [with unmasked proc mount type and no user namespace] Expected error type "Invalid value" with detail "`hostUsers` must be false to use `Unmasked`", got []
```

**Comment by haircommander** [comment]
/hold cancel

figured it out

File: pkg/features/kube_features.go (hunk 6 above)
Comment (liggitt) [reply pkg/features/kube_features.go]: Thinking briefly about this, I _think_ the following are sufficient (would be good to get a check of that from someone like @deads2k during PRR of the beta promotion):

* All API servers in the cluster have the validation added in this PR in place
* All API servers in the cluster prevent the feature-gate combination `UserNamespacesSupport=false,ProcMountType=true`

We can make the second change in 1.30. We can't go back in time and make 1.29 servers have this validation or enforce that feature-gate pairing.

**Pull request merged**